File: wfirst_phaseb/matlab/examples/run_spc_ifs.m
clear all
delete(gcp('nocreate'))

npsf = 256;

lambda0 = 0.730; 
nlam = 9;  
bw = 0.18;
lam_array = [(1-bw/2):bw/(nlam-mod(nlam,2)):(1+bw/2)]*lambda0;

% compute unberrated coronagraphic field 

optval.cor_type ='spc-ifs_long'; 
optval.final_sampling_lam0 = 0.1;
optval.use_errors = 0;
optval.zindex = 4;
optval.zval_m = 0.19e-9;

fields = prop_run_multi( ['wfirst_phaseb'], lam_array, npsf, 'quiet', 'passvalue',optval );
image  = sum(abs(fields).^2,3) / nlam;

% offaxis compact psf pk for normalizing factor (no errors) 

optval.zindex = 0;
optval.zval_m = 0;
optval.source_x_offset = 7.0;

fields = prop_run_multi( ['wfirst_phaseb_compact'], lam_array, npsf, 'quiet', 'passvalue',optval );
psf  = sum(abs(fields).^2,3) / nlam;
max_psf = max(max(psf));

ni = image / max_psf;

figure(1), clf, imagesc([log10(ni)]);
colorbar, caxis([-10 -7]), axis image
colormap jet

return

File: wfirst_phaseb/python/wfirst_phaseb_proper/examples/run_spc_ifs.py
#   Copyright 2019 California Institute of Technology
# ------------------------------------------------------------------

import proper
import numpy as np
import matplotlib.pylab as plt
from matplotlib.colors import LogNorm
from matplotlib.patches import Circle
from wfirst_phaseb_proper import trim

def run_spc_ifs():

    nlam = 9
    lam0 = 0.73
    bandwidth = 0.18
    lam_min = lam0 * (1 - bandwidth/2)
    lam_max = lam0 * (1 + bandwidth/2)
    lam = np.linspace( lam_min, lam_max, nlam )

    n = 256                 # output image dimension
    final_sampling = 0.1    # output sampling in lam0/D

    # compute coronagraphic image

    print( "Computing coronagraphic field..." )
    (fields, sampling) = proper.prop_run_multi('wfirst_phaseb', lam, n, QUIET=True, \
        PASSVALUE={'cor_type':'spc-ifs_long','lam0':lam0,'use_errors':0,'zindex':[4],'zval_m':[0.19e-9],'final_sampling_lam0':final_sampling} )
    images = np.abs(fields)**2
    image = np.sum( images, 0 ) / nlam

    # move source to 7 lam/D

    print( "Computing offset source to compute NI..." )
    (fields, sampling) = proper.prop_run_multi('wfirst_phaseb_compact', lam, n, QUIET=True, \
        PASSVALUE={'cor_type':'spc-ifs_long','lam0':lam0,'source_x_offset':7.0,'final_sampling_lam0':final_sampling} )
    psfs = np.abs(fields)**2
    psf = np.sum( psfs, 0 ) / nlam

    ni = image / np.max(psf)

    fig, ax  = plt.subplots( figsize=(8,8) )
    im = ax.imshow(ni, norm=LogNorm(vmin=1e-10,vmax=1e-7), cmap=plt.get_cmap('jet'))
    circ = Circle((n/2,n/2),3/final_sampling,edgecolor='white', facecolor='none')
    ax.add_patch(circ)
    circ = Circle((n/2,n/2),9/final_sampling,edgecolor='white', facecolor='none')
    ax.add_patch(circ)

    fig.colorbar(im, shrink=0.5) 
    plt.show()
    
if __name__ == '__main__':
    run_spc_ifs()
